### PR TITLE
Remove unstated and incorrect assumptions about CBF metadata for a multi-axis goniometer

### PR DESCRIPTION
**Pull request by benjaminhwilliams**
At present, there are two unstated (and incorrect) assumptions about the way goniometer axis metadata must be formatted when constructing a multi-axis diffractometer model from CBF image files.  The result is that perfectly valid data may be misinterpreted either because `diffrn_scan_axis` category of the metadata is ordered differently to the `axis` category, or because `axis` is not in order of axis dependency.

With this change, dictionaries are used to record the axis metadata and the axis names are used as keys (that's what they're there for, after all).  The order in which the metadata are listed in either category is immaterial.

<details>
<summary>
Details below of the two assumptions that this PR corrects.  Click to expand.
</summary>

## Ordering of the `axis` category
Firstly, it is assumed that goniometer axes in the `axis` category will be listed in order of dependency, starting with the axis mounted on the lab floor and ending with the axis on which the sample is mounted (e.g. the order _ω_, _κ_, _φ_ for a _κ_-geometry diffractometer).  This assumption is not evident from the code and seems to be unintentional.  Note first how the axis dependencies and vectors are read from the `axis` category of an image header:
https://github.com/cctbx/dxtbx/blob/927539e8cf40b58b9598dbef6149bcf710324081/src/dxtbx/model/goniometer.py#L400-L413
Then notice that there is an attempt to assemble the axes in dependency order:
https://github.com/cctbx/dxtbx/blob/927539e8cf40b58b9598dbef6149bcf710324081/src/dxtbx/model/goniometer.py#L440-L448
This reordering is flawed.  Take this hypothetical five-axis goniometer as an example, with axes named 'A'–'E', listed here in the order that they are recorded in the `axis` category of a CBF header, alongside the `order` deduced as above:
| Index | `axis_names` | `depends_on` | `order`
| --- | --- | --- | --- | 
| 0 | A | E | 5 |
| 1 | B | A | 1 |
| 2 | C | D | 4 |
| 3 | D | `"."` (i.e. the lab) | 0 |
| 4 | E | C | 3 |

Remember, we're going to use `order` as the sort index for `axis_names` (and some other sequences, `axes` and `angles`, which have the same length).  So where did the index 2 go?  And how on earth are we to interpret an index of 5 in a five-membered sequence?

Note that the ordering _does_ work iff the axes happen already to be listed in dependency order:
| Index | `axis_names` | `depends_on` | `order`
| --- | --- | --- | --- | 
| 0 | A | `"."` (i.e. the lab) | 0 |
| 1 | B | A | 1 |
| 2 | C | B | 2 |
| 3 | D | C | 3 |
| 4 | E | D | 4 |

So in the only case where the ordering works, the ordering is an elaborate no-op.  :shrug:

## Ordering of the `diffrn_scan_axis` category
Separately, there is an assumption that the rotation angle metadata in the `diffrn_scan_axis` category will list the axes in the same order as in the `axis` category.  The rotation `angles` are simply read in order from the `diffrn_scan_axis` block, just as the axis vectors `axis` and dependencies `depends_on` are read in the order that they appear in the `axis` block:
https://github.com/cctbx/dxtbx/blob/927539e8cf40b58b9598dbef6149bcf710324081/src/dxtbx/model/goniometer.py#L417-L428
Then the `order` deduced above (leaving aside that it is flawed unless it is essentially simply `range(len(axis_names))`) is used to reorder `angles`, just as for the axis vectors `axes` and the `axis_names` from the `axis` metadata category:
https://github.com/cctbx/dxtbx/blob/927539e8cf40b58b9598dbef6149bcf710324081/src/dxtbx/model/goniometer.py#L453-L455

If the CBF headers have the `diffrn_scan_axis` category ordered differently to the `axis` category, then the start angles for some axes will be assigned erroneously to others.  🤦
</details>

### CODE REVIEW

**Comment by benjaminhwilliams** [comment]
I think I have a lightweight idea for how to knock together some simple tests for this.  I'll add that when I get a chance.

File: src/dxtbx/model/goniometer.py [hunk 1]
```diff
@@ -392,69 +392,69 @@ def imgCIF_H(cbf_handle):
...
+            ordered_axes.append(axis)
 
         # multi-axis gonio requires axes in order as viewed from crystal to gonio base
         # i.e. the reverse of the order we have from cbf header
-        order = order.reversed()
-        axes = axes.select(order)
-        angles = angles.select(order)
-        axis_names = axis_names.select(order)
-        scan_axis = axes.size() - scan_axis - 1
+        ordered_axes = ordered_axes[::-1]
```
Comment (graeme-winter) [inline src/dxtbx/model/goniometer.py:450]: This is preferable to `reversed()`?

File: newsfragments/539.bugfix [hunk 2]
```diff
@@ -0,0 +1,6 @@
+This removes two unstated assumptions about the way goniometer axis metadata must be formatted when constructing a multi-axis diffractometer model from CBF image files.
+Firstly, it had been assumed that goniometer axes in the ``axis`` category would be listed in order of dependency, starting with the axis mounted on the lab floor and ending with the axis on which the sample is mounted (e.g. the order *ω*, *κ*, *φ* for a *κ*-geometry diffractometer).
+Separately, there was an assumption that the rotation angle metadata in the ``diffrn_scan_axis`` category would list the axes in the same order as in the ``axis`` category.
+The result was that perfectly valid data would be misinterpreted either because ``diffrn_scan_axis`` was ordered differently to ``axis``, or because the ``axis`` metadata were not in order of axis dependency.
```
Comment (graeme-winter) [inline newsfragments/539.bugfix:4]: s/would/may be misinterpreted

File: src/dxtbx/model/goniometer.py [hunk 3]
```diff
@@ -392,69 +392,69 @@ def imgCIF_H(cbf_handle):
...
-            angles.append(axis_angle)
-            if abs(increment) > 0:
-                assert (
-                    scan_axis is None
-                ), "More than one scan axis is defined: not currently supported"
-                scan_axis = flex.first_index(axis_names, axis_name)
+            angles[axis_name] = axis_angle
+            if increment != 0:
+                if scan_axis:
+                    raise ValueError(
```
Comment (graeme-winter) [inline src/dxtbx/model/goniometer.py:430]: Should we try and fake some data to exercise this?

File: src/dxtbx/model/goniometer.py [hunk 4]
```diff
@@ -392,69 +392,69 @@ def imgCIF_H(cbf_handle):
...
+                        "More than one scan axis is defined: not currently supported."
+                    )
+                scan_axis = axis_name
             cbf_handle.next_row()
-        assert axes.size() == angles.size()
-        if scan_axis is None:
-            # probably a still shot -> scan axis arbitrary as no scan
-            scan_axis = 0
+        if not len(axis_vectors) == len(angles):
+            raise ValueError(
```
Comment (graeme-winter) [inline src/dxtbx/model/goniometer.py:436]: Should we try and fake some data to exercise this?

File: src/dxtbx/model/goniometer.py [hunk 5]
```diff
@@ -392,69 +392,69 @@ def imgCIF_H(cbf_handle):
...
-        for i in range(axes.size()):
-            if depends_on[i] == ".":
-                o = 0
-            else:
-                o = flex.first_index(axis_names, depends_on[i]) + 1
-            assert o not in order
-            order.append(o)
+        ordered_axes = []
+        axis = "."
+        while axis := dependants.get(axis):
```
Comment (graeme-winter) [inline src/dxtbx/model/goniometer.py:445]: Ohai walrus -> be aware when we backport

**Review by graeme-winter — APPROVED** [review]
Important and well implemented set of changes thank you - a handful of small comments but nothing blocker like. 

Happy to help try and cook up some examples to test the failure modes.

File: src/dxtbx/model/goniometer.py (hunk 1 above)
Comment (benjaminhwilliams) [reply src/dxtbx/model/goniometer.py:450]: Reversed returns an iterator, so you can't use it twice.

Comment (graeme-winter) [reply src/dxtbx/model/goniometer.py:450]: 👍 so yes it is

File: newsfragments/539.bugfix (hunk 2 above)
Comment (benjaminhwilliams) [reply newsfragments/539.bugfix:4]: In the changelog, I'm referring to the issue in the past tense, so perhaps `s/would be/may have been/`.  Except that I'm specifically describing the cases where the misinterpretation _would_ happen, so there's no uncertainty about it.

Comment (benjaminhwilliams) [reply newsfragments/539.bugfix:4]: Ah, but I would agree with `s/because/when/g`.

File: src/dxtbx/model/goniometer.py (hunk 5 above)
Comment (benjaminhwilliams) [reply src/dxtbx/model/goniometer.py:445]: Ah yes, of course.  :roll_eyes:

File: src/dxtbx/model/goniometer.py (hunk 3 above)
Comment (benjaminhwilliams) [reply src/dxtbx/model/goniometer.py:430]: Exactly my plan.

**Comment by benjaminhwilliams** [comment]
Since @graeme-winter has reviewed, I'll remove the request for @rjgildea to review.  Of course, a review is still welcome if you want to leave one, but don't feel obliged.

**Comment by benjaminhwilliams** [comment]
It seems making tests for this will take a little time because we don't have any three-axis goniometer CBF data uploaded to anywhere like Zenodo.  The closest we have is [some two-axis goniometer data](https://zenodo.org/record/51405#.YvevwtLMJH4) from I19-1 at Diamond Light Source, which might be enough, but which doesn't yet have a record in [dials/data](https://github.com/dials/data).

Given that the untested behaviour was untested before this PR, perhaps we can decouple the fix from the gap in the test coverage by merging this now and addig the tests as a separate change set?

**Comment by graeme-winter** [comment]
> It seems making tests for this will take a little time because we don't have any three-axis goniometer CBF data uploaded to anywhere like Zenodo. The closest we have is [some two-axis goniometer data](https://zenodo.org/record/51405#.YvevwtLMJH4) from I19-1 at Diamond Light Source, which might be enough, but which doesn't yet have a record in [dials/data](https://github.com/dials/data).
> 
> Given that the untested behaviour was untested before this PR, perhaps we can decouple the fix from the gap in the test coverage by merging this now and addig the tests as a separate change set?

100% support this approach

**Pull request merged**